### ISSUE TITLE
Stopping Criteria

### ISSUE THREAD

**Issue by hakan-gecili** [ljvmiranda921/pyswarms#373]
* PySwarms version: All version
* Python version: Python 3
* Operating System: All OS

### Description
User should be able to define the stopping criteria as " If the objective function value does not improve for N iterations stop the optimization and return the best objective function value and other related objects". 

This feature has been widely used in metaheuristic optimization and it is very simple to implement. I checked the code and documentation and haven't seen this feature. There are 3 different stop criteria defined ( maxiter , minstep, minfunc), none of these can achieve what I described in the first paragraph.

### What I Did
I am happy to help if needed. 
Thank you.

**Comment by whzup** [comment]
Hello @hakan-gecili, and welcome at PySwarms! How would you imagine this feature to look like? A parameter for tolerance `tol` and one for steps `n` and the condition will then be

"If the value of the last `n` steps are all within the tolerance `tol`, then stop"?

**Comment by hakan-gecili** [comment]
Hi Aaron,

Your description is good but to make sure that we are on the same page.
I would define *tol* as the tolerance that will be compared with the
difference in objective function values (OFV) of consecutive iterations or
steps.
To be more specific:

delta_list = [None] * n
previous_best_cost = # run the PSO here and get an initial solution cost
for t in range(1, nb_iterations)
    new_best_cost = # run the PSO here and get a best_cost
    delta = new_best_cost - previous_best_cost
    if t < n:
        delta_list[t] = delta
    else:
        delta_list = delta_list[1:] + [delta]
        if all([j < tol for j in delta_list]):
            # Converged, terminate the optimization
            return (best_cost, best_pos)
    previous_best_cost = new_best_cost

I am working at Wright State University and we solve large scale
optimization problems, for those problems it is not possible to guess the
number of iterations, so we use convergence as a stopping criteria.

Please let me know if you have any questions

Thank you

**Comment by hakan-gecili** [comment]
``delta_list = [None] * n
previous_best_cost = # run the PSO here and get an initial solution cost
for t in range(1, nb_iterations)
    new_best_cost = # run the PSO here and get a best_cost
    delta = new_best_cost - previous_best_cost
    if t < n:
        delta_list[t] = delta
    else:
        delta_list = delta_list[1:] + [delta]
        if all([j < tol for j in delta_list]):
            # Converged, terminate the optimization
            return (best_cost, best_pos)
    previous_best_cost = new_best_cost
``

**Comment by hakan-gecili** [comment]
``

**Comment by hakan-gecili** [comment]
Hi Aaron,

Your description is good but to make sure that we are on the same page.
I would define *tol* as the tolerance that will be compared with the
difference in objective function values (OFV) of consecutive iterations or
steps.
To be more specific:

```python
delta_list = [None] * n
previous_best_cost = # run the PSO here and get an initial solution cost
for t in range(1, nb_iterations)
    new_best_cost = # run the PSO here and get a best_cost
    delta = new_best_cost - previous_best_cost
    if t < n:
        delta_list[t] = delta
    else:
        delta_list = delta_list[1:] + [delta]
        if all([j < tol for j in delta_list]):
            # Converged, terminate the optimization
            return (best_cost, best_pos)
    previous_best_cost = new_best_cost
```

I am working at Wright State University and we solve large scale
optimization problems, for those problems it is not possible to guess the
number of iterations, so we use convergence as a stopping criteria.

Please let me know if you have any questions

Thank you

**Comment by whzup** [comment]
@hakan-gecili ok, so this would then be kind of an extension of the `ftol` option we already have. Only that instead of only 2 values we look at `n` values, right?

**Comment by hakan-gecili** [comment]
Correct

**Comment by ewekam** [comment]
Hi,
Is there a chance that this enhacement will be implemented in some near future? 
In my project, we would like to to use this stopping criteria as well. 
Thanks, 
Ewelina

**Comment by whzup** [comment]
Hello @ewekam, would you be interested in creating a PR for this purpose? or you @hakan-gecili? Unfortunately, at the moment, I have a lot to do with my thesis 😕.

**Comment by whzup** [comment]
@hakan-gecili It seems like @nishnash54 has already taken on this issue! 🎊 Thanks a lot! 👍

**Comment by ljvmiranda921** [comment]
Closing this issue now since it's resolved by #391 
Will make a new release with this feature by end of month
Thanks everyone for the help!
